Transform CFGs into nested form

acl-cqc opened the pull request:
* Rename CfgView -> CfgNodeMap, and SimpleCfgView -> IdentityNodeMap
* Add CfgNester subtrait with a method to perform a single nesting transform
* Change CfgNodeMap + HalfNodeView to take the <H:HugrView/Mut> rather than an `&'a (mut) H`
* Implement CfgNester for IdentityNodeMap using OutlineCfg rewrite (no impl for HalfNodeView yet)
* Add transform_cfg_to_nested that recursively traverses any CfgNodeMap and nests all regions discovered in the CFG, innermost-first
* And transform_all_cfgs that runs the previous with an IdentityNodeMap for each CFG found in the entire Hugr

acl-cqc commented:
@mark-koch this is not really "for review" - but, please take a look, and see if you can use it?

mark-koch commented:
I made a Guppy branch to play around with this: https://github.com/CQCL-DEV/guppy/tree/toy/nesting

To set everything up, checkout the branch and run
```sh
python -m venv .venv
source .venv/bin/activate
pip install -e '.[dev]'
cd validator
maturin develop
cd ..
```

Then you can play around with the file `nesting.py`. The current example in the file produces a failed assertion, but maybe I'm using it wrong?

acl-cqc commented:
> The current example in the file produces a failed assertion, but maybe I'm using it wrong?

As per description:

> all the nest_cfgs.rs analysis works only on whole CFG-rooted Hugrs (!)

This is a temporary measure until we solve https://github.com/CQCL-DEV/hugr/issues/246, but I could easily add a parameter to the CFG node to analyze.

acl-cqc commented:
@mark-koch I've added a new function `transform_all_cfgs` that we can play around with for the time being, but this is probably not a high priority

acl-cqc commented on src/algorithm/nest_cfgs.rs:
```diff
@@ -43,21 +43,23 @@ use std::hash::Hash;
+/// algorithm to identify more regions than existed in the underlying CFG
 /// (without mutating the underlying CFG perhaps in vain).
-pub trait CfgView<T> {
+pub trait CfgNodeMap<T> {
```
I'm renaming this as a could-be-separate part of the PR ;). While I'm about it, I could make the `T` a `type` member of the trait (`NodeType`, say)...

acl-cqc commented on src/algorithm/nest_cfgs.rs:
```diff
@@ -71,6 +74,92 @@ pub trait CfgView<T> {
+    }
+
+    // Traverse. Any traversal will encounter edges in SESE-respecting order.
+    fn traverse<T: Copy + Eq + Hash + std::fmt::Debug>(
```
This has to be recursive because the calls with the `stop_at` parameter are too complex to use the internal `node_stack`. I tried using [fix_fn](https://crates.io/crates/fix_fn) to make a recursive closure (which would avoid passing `edge_classes` and `rem_edges`) but that doesn't allow declaring type parameters or using them from the containing method :-(

acl-cqc commented on src/algorithm/nest_cfgs.rs:
```diff
@@ -71,6 +75,92 @@ pub trait CfgView<T> {
+    ///
+    /// # Panics
+    /// May panic if the two edges do not constitute a SESE region.
+    fn nest_sese_region(&mut self, entry_edge: (T, T), exit_edge: (T, T)) -> T;
```
It is possible we should return some kind of Result, but it's not really clear what. The algorithm here only calls this for single-entry-single-exit regions identified by the analysis, so this should not fail. Perhaps a more complex splitting/merging CfgNodeMap might want to fail in some cases - but *that* I think we should leave for when a fuller splitting/merging CfgNodeMap is developed, i.e. not this PR.

acl-cqc commented on src/algorithm/nest_cfgs.rs:
```diff
@@ -445,14 +620,14 @@ pub(crate) mod test {
-        // that we *can* (as we'll need to for "real" module Hugr's).
-        let v: SiblingGraph = SiblingGraph::try_new(&h, h.root()).unwrap();
-        let edge_classes = EdgeClassifier::get_edge_classes(&SimpleCfgView::new(&v));
+        // ALAN here's where we need a WholeHugrView...tho it's good to try SiblingMut at least somewhere
```
This refers to the WholeHugrView in #548 .... however, even if that goes in, I'm not sure whether it'd be good to use it or not now, given the traversal algorithm in `transform_cfg_to_nested` has been specifically designed to work on a `SiblingMut`. So I should either use WholeHugrView; or stick with SiblingMut and drop this comment, and the other mentions.

cqc-alec reviewed (commented):
Looks good, just small comments. Would be good to have some more complicated test cases soon.

acl-cqc commented on src/algorithm/half_node.rs:
```diff
@@ -110,8 +119,13 @@ mod test {
-        let v = HalfNodeView::new(&h);
+        // Allowing to identity two nested regions (and fixing the problem with an IdentityCfgMap on the same example)
+
+        // ALAN and another case for WholeHugrView
```
Yes, I think it does not belong here. I thought that, since the RootHandle tag-checking had been added to main *since* this PR started, I could roll into this PR updates to the existing structure-detection code to use said tag-checking....but that leads to wanting #548 which has ended up significantly more complicated/longer-lived than expected, so I think it's best to leave this for #548 or another PR. I don't have an issue but I can roll it into that PR easily if this goes in first.

acl-cqc commented on src/algorithm/nest_cfgs.rs:
```diff
@@ -42,22 +42,26 @@ use std::collections::{HashMap, HashSet, LinkedList, VecDeque};
+/// for an [IdentityCfgMap] if the extra level of indirection is not required. However, since
+/// SESE regions are bounded by edges between pairs of such `T`, such splitting may allow the
+/// algorithm to identify more regions than existed in the underlying CFG, without mutating the
+/// underlying CFG just for the analysis - the splitting /merging) can then be performed by
```
I've gone with "splitting and/or merging"

acl-cqc commented on src/algorithm/half_node.rs:
```diff
@@ -87,14 +90,20 @@ impl<H: HugrView> CfgView<HalfNode> for HalfNodeView<'_, H> {
     }
+
+    fn nest_sese_region(&mut self, _i: (HalfNode, HalfNode), _x: (HalfNode, HalfNode)) -> HalfNode {
+        todo!()
```
There is now, #593. (Previously the issue was only on JIRA and there is more discussion linked from there.)

acl-cqc commented:
Ok, thanks for the prodding :), I've changed a few things around...
* Separate the `nest_sese_region` function out into a different trait. This allows doing the analysis (not the transformation) using a read-only HugrView, which seems useful.
   * This required IdentityCfgMap (and HalfNodeView) to take `H: HugrView`s rather than references thereof, as you can't parameterize over "whether a reference is &mut or &". Region views are cloneable, *references* to Hugrs are trivially too, so this seems ok.
   * The change to trait structure also leads to the HalfNode transform disappearing entirely (not even a `todo!()`), but there's now an issue, as you suggested
* Dropped all references to WholeHugrView, enabling root-node-tagging is a separate issue, doesn't need to be here

cqc-alec commented on src/algorithm/half_node.rs:
```diff
@@ -110,8 +114,10 @@ mod test {
         //               |          \-> right -/                 |
         //               \---<---<---<---<---<---<---<---<---<---/
-        // Allowing to identity two nested regions (and fixing the problem with a SimpleCfgView on the same example)
+        // Allowing to identity two nested regions (and fixing the problem with an IdentityCfgMap on the same example)
```
```suggestion
        // Allowing to identify two nested regions (and fixing the problem with an IdentityCfgMap on the same example)
```
(Not strictly part of this PR but may as well fix here.)
